lib/ecto/repo/schema.ex

```diff
@@ -932,6 +932,19 @@ defmodule Ecto.Repo.Schema do
     end
   end
 
+  defp dump_field!(action, schema, field, {:parameterized, type, opts}, value, adapter) do
+    base_type = type.type(opts)
+    with {:ok, parameterized_value} <- type.dump(value, opts),
+         {:ok, final_value} <- Ecto.Type.adapter_dump(adapter, base_type, parameterized_value)     do
+      final_value
+    else
+      :error ->
+        raise Ecto.ChangeError,
+              "value `#{inspect(value)}` for `#{inspect(schema)}.#{field}` " <>
+              "in `#{action}` does not match type #{type}"
+    end
+  end
```

Comment:
Hrm... is the issue here that the result of the parameterized dump still needs to be dumped? 🤔 I wonder if we should move this to a feature set of the parameterized type itself but I am not sure how. Either `dump` can return a special tuple that means it needs to be further dumped or we pass a callback to `dump` so it happens recursively.

Reply:
Maybe I am approaching it wrong, or misunderstood the dump function, but as it stands, I modeled parameterized types following the pattern of the `type` callback on custom types, they still have an underlying "base" data structure such as a map or string. Parameterized types know how to get themselves into their own base format (e.g. Embed "dumps" to a :map), and then Ecto knows how to "dump" the base type. The reasoning on this would be that authors of parameterized types would only have to worry about dumping their data to a higher level type like map, and ecto can then handle the mechanics of taking it from a map to the underlying data structure. They could always specify a type of :any and return the base storable values, as that would basically make the second call in the with statement a noop.

Reply:
What you said makes perfect sense. So I believe what needs to be "fixed" here is to move this logic to inside Ecto.Type, rather than here. :)